src/ray/gcs/gcs_server/gcs_table_storage.cc

```diff
@@ -84,8 +95,11 @@ Status GcsTableWithJobId<Key, Data>::Put(const Key &key,
     absl::MutexLock lock(&mutex_);
     index_[GetJobIdFromKey(key)].insert(key);
   }
```

Comment:
I actually have some questions about this one. Here the writing is async but the updating of index is sync. Will this cause some issues? I mean for example, if the put failed or the put is in the fly, but the user can see the index there.

Reply:
That's right, there will be the case where the index is created by the value is not in the storage. The correctness is guaranteed by always query the storage when doing index based query (in GcsTableWithJobId<Key, Data>::GetByJobId). So in this particular case even the index returns a key the GetByJobId will still return empty result.

Reply:
Got it. Please put some comment in the code about this.

Reply:
Another question is about the delete. Sorry I should catch this in the previous PR. But here, check the code:

```
template <typename Key, typename Data>
Status GcsTableWithJobId<Key, Data>::DeleteByJobId(const JobID &job_id,
                                                   const StatusCallback &callback) {
  std::vector<Key> keys;
  {
    absl::MutexLock lock(&mutex_);
    auto &key_set = index_[job_id];
    for (auto &key : key_set) {
      keys.push_back(key);
    }
  }
  return BatchDelete(keys, callback);
}
```

Could there be a race condition where we have the key but are not indexed?

1. key written into the index
2. key got deleted by calling DeleteByJobId
3. key got written into db.

Reply:
yup it will be problematic. I believe this is already a bug with redis based version so we are not worse here.
To solve it we need to serialize all the writes to `GcsTableWithJobId`